### PR TITLE
Remote enforcer kill is done in parallel now. Also log number of remote enforcers left to exit. 

### PR DESCRIPTION
**Pull request by 0xmchadha**

### CODE REVIEW

**Comment by 0xmchadha** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "",
    "component": "enforcerd",
    "pr-id": "1614",
    "commit-sha": "b3db3151dd02624539438a553c9914e5e2cfb13a",
    "pipeline": "master"
  },
  {
    "project": "",
    "component": "trireme-lib",
    "pr-id": "994",
    "commit-sha": "8982a0c3affe3c4ddc1f419d9f7930ca7e122f7e",
    "pipeline": "master"
  }
]
```

**Comment by 0xmchadha** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "",
    "component": "enforcerd",
    "pr-id": "1614",
    "commit-sha": "b3db3151dd02624539438a553c9914e5e2cfb13a",
    "pipeline": "master"
  },
  {
    "project": "",
    "component": "trireme-lib",
    "pr-id": "994",
    "commit-sha": "77a3c523691e2efb7c7052f2b09de50b6a52c755",
    "pipeline": "master"
  }
]
```

**Comment by 0xmchadha** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "",
    "component": "enforcerd",
    "pr-id": "1614",
    "commit-sha": "b3db3151dd02624539438a553c9914e5e2cfb13a",
    "pipeline": "master"
  },
  {
    "project": "",
    "component": "trireme-lib",
    "pr-id": "994",
    "commit-sha": "4318eca8b607f34cbd34fb95dce0896af2ea12a1",
    "pipeline": "master"
  }
]
```

**Comment by 0xmchadha** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "",
    "component": "enforcerd",
    "pr-id": "1614",
    "commit-sha": "b3db3151dd02624539438a553c9914e5e2cfb13a",
    "pipeline": "master"
  },
  {
    "project": "",
    "component": "trireme-lib",
    "pr-id": "994",
    "commit-sha": "4318eca8b607f34cbd34fb95dce0896af2ea12a1",
    "pipeline": "master"
  }
]
```

File: controller/internal/enforcer/proxy/enforcerproxy.go [hunk 1]
```diff
@@ -169,27 +188,66 @@ func (s *ProxyInfo) SetLogLevel(level constants.LogLevel) error {
...
 func (s *ProxyInfo) CleanUp() error {
+	var contextIDs []string
+	var synch sync.Mutex
+	var wg sync.WaitGroup
 
-	var allErrors string
+	s.Lock()
 
-	remotes := s.rpchdl.ContextList()
+	for contextID, _ := range s.remoteEnforcers {
```
Comment (dstiliadis) [inline controller/internal/enforcer/proxy/enforcerproxy.go]: The list of remotes is already maintained in rpchdl .. adding one more map to maintain the list is possible to create problesm. Especially in the case of remote failures/restarts ... Try it out where you kill some remotes (or make them fail in a loop).

File: controller/internal/enforcer/proxy/enforcerproxy.go [hunk 2]
```diff
@@ -169,27 +188,66 @@ func (s *ProxyInfo) SetLogLevel(level constants.LogLevel) error {
...
+			for {
+				synch.Lock()
+				if len(contextIDs) == 0 {
+					synch.Unlock()
+					return
+				}
+				cid = contextIDs[0]
+				contextIDs = contextIDs[1:]
+				synch.Unlock()
+
```
Comment (dstiliadis) [inline controller/internal/enforcer/proxy/enforcerproxy.go]: I think this is better done with a buffered channel rather than slices in go.

**Comment by 0xmchadha** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "",
    "component": "enforcerd",
    "pr-id": "1614",
    "commit-sha": "b3db3151dd02624539438a553c9914e5e2cfb13a",
    "pipeline": "master"
  },
  {
    "project": "",
    "component": "trireme-lib",
    "pr-id": "994",
    "commit-sha": "4318eca8b607f34cbd34fb95dce0896af2ea12a1",
    "pipeline": "master"
  }
]
```

**Comment by 0xmchadha** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "",
    "component": "enforcerd",
    "pr-id": "1614",
    "commit-sha": "4e9fa06361c1cba840be58a4863f05dfd793f4fe",
    "pipeline": "master"
  },
  {
    "project": "",
    "component": "trireme-lib",
    "pr-id": "994",
    "commit-sha": "cf8887892ebc573f6e09828ebe08c30f2ffb8802",
    "pipeline": "master"
  }
]
```

**Comment by 0xmchadha** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "",
    "component": "enforcerd",
    "pr-id": "1614",
    "commit-sha": "4e9fa06361c1cba840be58a4863f05dfd793f4fe",
    "pipeline": "master"
  },
  {
    "project": "",
    "component": "trireme-lib",
    "pr-id": "994",
    "commit-sha": "cf8887892ebc573f6e09828ebe08c30f2ffb8802",
    "pipeline": "master"
  }
]
```

**Comment by 0xmchadha** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "",
    "component": "enforcerd",
    "pr-id": "1614",
    "commit-sha": "4e9fa06361c1cba840be58a4863f05dfd793f4fe",
    "pipeline": "master"
  },
  {
    "project": "",
    "component": "trireme-lib",
    "pr-id": "994",
    "commit-sha": "cf8887892ebc573f6e09828ebe08c30f2ffb8802",
    "pipeline": "master"
  }
]
```

**Comment by 0xmchadha** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "",
    "component": "enforcerd",
    "pr-id": "1614",
    "commit-sha": "4e9fa06361c1cba840be58a4863f05dfd793f4fe",
    "pipeline": "master"
  },
  {
    "project": "",
    "component": "trireme-lib",
    "pr-id": "994",
    "commit-sha": "346f237d77f4860efa44daafd06be62f43a212b3",
    "pipeline": "master"
  }
]
```

**Comment by 0xmchadha** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "",
    "component": "enforcerd",
    "pr-id": "1614",
    "commit-sha": "4e9fa06361c1cba840be58a4863f05dfd793f4fe",
    "pipeline": "master"
  },
  {
    "project": "",
    "component": "trireme-lib",
    "pr-id": "994",
    "commit-sha": "346f237d77f4860efa44daafd06be62f43a212b3",
    "pipeline": "master"
  }
]
```

**Comment by 0xmchadha** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "",
    "component": "enforcerd",
    "pr-id": "1614",
    "commit-sha": "4e9fa06361c1cba840be58a4863f05dfd793f4fe",
    "pipeline": "master"
  },
  {
    "project": "",
    "component": "trireme-lib",
    "pr-id": "994",
    "commit-sha": "a231cb217971c90c16e9e4347266931ad3e28952",
    "pipeline": "master"
  }
]
```

**Comment by 0xmchadha** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "",
    "component": "enforcerd",
    "pr-id": "1614",
    "commit-sha": "4e9fa06361c1cba840be58a4863f05dfd793f4fe",
    "pipeline": "master"
  },
  {
    "project": "",
    "component": "trireme-lib",
    "pr-id": "994",
    "commit-sha": "eaae0a24cb4a0531c44d196f33fe487bba9672d5",
    "pipeline": "master"
  }
]
```

**Comment by 0xmchadha** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "",
    "component": "enforcerd",
    "pr-id": "1614",
    "commit-sha": "4e9fa06361c1cba840be58a4863f05dfd793f4fe",
    "pipeline": "master"
  },
  {
    "project": "",
    "component": "trireme-lib",
    "pr-id": "994",
    "commit-sha": "eaae0a24cb4a0531c44d196f33fe487bba9672d5",
    "pipeline": "master"
  }
]
```

**Comment by 0xmchadha** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "",
    "component": "enforcerd",
    "pr-id": "1614",
    "commit-sha": "4e9fa06361c1cba840be58a4863f05dfd793f4fe",
    "pipeline": "master"
  },
  {
    "project": "",
    "component": "trireme-lib",
    "pr-id": "994",
    "commit-sha": "eaae0a24cb4a0531c44d196f33fe487bba9672d5",
    "pipeline": "master"
  }
]
```

**Comment by 0xmchadha** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "",
    "component": "enforcerd",
    "pr-id": "1614",
    "commit-sha": "4e9fa06361c1cba840be58a4863f05dfd793f4fe",
    "pipeline": "master"
  },
  {
    "project": "",
    "component": "trireme-lib",
    "pr-id": "994",
    "commit-sha": "eaae0a24cb4a0531c44d196f33fe487bba9672d5",
    "pipeline": "master"
  }
]
```

**Comment by 0xmchadha** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "",
    "component": "enforcerd",
    "pr-id": "1614",
    "commit-sha": "3e7abfa9447f6ca06a76e894ba292a4c212657e2",
    "pipeline": "master"
  },
  {
    "project": "",
    "component": "trireme-lib",
    "pr-id": "994",
    "commit-sha": "2b87aa97f52361a0cac4387f20a5c6a184ae962d",
    "pipeline": "master"
  }
]
```

**Review by amitlimaye — APPROVED** [review]

**Pull request merged**